homeassistant/components/mazda/climate.py

```diff
@@ -0,0 +1,186 @@
+
+        self.region = region
+        self._attr_unique_id = self.vin
+
+        if self.data["hvacSetting"]["temperatureUnit"] == "F":
+            self._attr_precision = PRECISION_WHOLE
+            self._attr_temperature_unit = TEMP_FAHRENHEIT
+            self._attr_min_temp = 61.0
+            self._attr_max_temp = 83.0
+        else:
+            self._attr_precision = PRECISION_HALVES
+            self._attr_temperature_unit = TEMP_CELSIUS
+            if region == "MJO":
+                self._attr_min_temp = 18.5
+                self._attr_max_temp = 31.5
+            else:
+                self._attr_min_temp = 15.5
+                self._attr_max_temp = 28.5
+
+        self._update_state_attributes()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Update attributes when the coordinator data updates."""
+        self._update_state_attributes()
+
+        super()._handle_coordinator_update()
+
+    def _update_state_attributes(self) -> None:
+        # Update the HVAC mode
+        hvac_on = self.client.get_assumed_hvac_mode(self.vehicle_id)
+        self._attr_hvac_mode = HVACMode.HEAT_COOL if hvac_on else HVACMode.OFF
+
+        # Update the target temperature
+        hvac_setting = self.client.get_assumed_hvac_setting(self.vehicle_id)
+        self._attr_target_temperature = hvac_setting.get("temperature")
+
+        # Update the current temperature
+        current_temperature_celsius = self.data["evStatus"]["hvacInfo"][
+            "interiorTemperatureCelsius"
+        ]
+        if self.data["hvacSetting"]["temperatureUnit"] == "F":
+            self._attr_current_temperature = convert_temperature(
+                current_temperature_celsius, TEMP_CELSIUS, TEMP_FAHRENHEIT
+            )
+        else:
+            self._attr_current_temperature = current_temperature_celsius
+
+        # Update the preset mode based on the state of the front and rear defrosters
+        front_defroster = hvac_setting.get("frontDefroster")
+        rear_defroster = hvac_setting.get("rearDefroster")
+        if front_defroster and rear_defroster:
+            self._attr_preset_mode = PRESET_DEFROSTER_FRONT_AND_REAR
+        elif front_defroster:
+            self._attr_preset_mode = PRESET_DEFROSTER_FRONT
+        elif rear_defroster:
+            self._attr_preset_mode = PRESET_DEFROSTER_REAR
+        else:
+            self._attr_preset_mode = PRESET_DEFROSTER_OFF
+
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set a new HVAC mode."""
+        if hvac_mode == HVACMode.HEAT_COOL:
+            await self.client.turn_on_hvac(self.vehicle_id)
+        elif hvac_mode == HVACMode.OFF:
+            await self.client.turn_off_hvac(self.vehicle_id)
+
+        self._handle_coordinator_update()
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
+        """Set a new target temperature."""
+        if (temperature := kwargs.get(ATTR_TEMPERATURE)) is not None:
```

Comment:
This would be cleaner if you reversed the condition and returned early so you could outdent below

Reply:
I actually did it this way because I couldn't find a way to achieve 100% test coverage otherwise. It doesn't seem possible to create a test scenario where `kwargs.get(ATTR_TEMPERATURE)` is `None`.

Reply:
We can make a service call that passes low and high target temps instead of `ATTR_TEMPERATURE`.